src/main/java/org/spongepowered/common/text/selector/SelectorResolver.java

```diff
@@ -324,64 +312,57 @@ private void addTypeFilters(List<Predicate<Entity>> filters) {
             Argument.Invertible<EntityType> typeArg = typeOpt.get();
             final boolean inverted = typeArg.isInverted();
             final EntityType type = typeArg.getValue();
-            filters.add(input -> inverted ^ input.getType() == type);
+            filters.add(input -> inverted != (input.getType() == type));
         }
     }
 
     private Vector3d getPositionOrDefault(Vector3d pos, ArgumentHolder.Vector3<?, ? extends Number> vecTypes) {
         Optional<Double> x = this.selector.get(vecTypes.x()).map(TO_DOUBLE);
         Optional<Double> y = this.selector.get(vecTypes.y()).map(TO_DOUBLE);
         Optional<Double> z = this.selector.get(vecTypes.z()).map(TO_DOUBLE);
-        return new Vector3d(x.orElse(Double.valueOf(pos.getX())), y.orElse(Double.valueOf(pos.getY())), z.orElse(Double.valueOf(pos.getZ())));
+        return new Vector3d(x.orElse(pos.getX()), y.orElse(pos.getY()), z.orElse(pos.getZ()));
     }
 
     public String getName() {
         return this.original.map(GET_NAME).orElse("SelectorResolver");
     }
 
-    public Set<Entity> resolve() {
+    public List<Entity> resolve() {
         SelectorType selectorType = this.selector.getType();
         int defaultCount = 1;
         if (INFINITE_TYPES.contains(selectorType)) {
             defaultCount = 0;
         }
         int maxToSelect = this.selector.get(ArgumentTypes.COUNT).orElse(defaultCount);
+        boolean isReversed = maxToSelect < 0;
+        maxToSelect = Math.abs(maxToSelect);
         Set<? extends Extent> extents = getExtentSet();
-        int count = 0;
-        ImmutableSet.Builder<Entity> entities = ImmutableSet.builder();
-        for (Extent extent : extents) {
-            Collection<Entity> allEntities = extent.getEntities();
+        Stream<Entity> entityStream = extents.stream()
+                .flatMap(ext -> ext.getEntities().stream())
+                .filter(this.selectorFilter);
+        if (maxToSelect != 0) {
             if (selectorType == SelectorTypes.RANDOM) {
-                List<Entity> entityList = new ArrayList<>(allEntities);
-                Collections.shuffle(entityList);
-                allEntities = entityList;
-            }
-
-            for (Entity e : allEntities) {
-                if (!this.selectorFilter.test(e)) {
-                    continue;
-                }
-                entities.add(e);
-                count++;
-                if (maxToSelect != 0 && count > maxToSelect) {
-                    break;
-                }
+                List<Entity> holder = entityStream.collect(Collectors.toList());
+                Collections.shuffle(holder);
```

Comment:
The shuffle won't have much effect because the stream will get sorted below. I am not sure what the right behavior would be, but are the entities always supposed to be sorted by distance? It also seems a bit waste of performance if you don't actually use it.